lib/repositories/chat_repository.dart

```diff
@@ -1,100 +1,98 @@
 
 import 'package:cloud_firestore/cloud_firestore.dart';
-import 'package:flutter_dotenv/flutter_dotenv.dart';
+import 'package:flutter/foundation.dart';
+import 'package:pass_emploi_app/models/conseiller_messages_info.dart';
 import 'package:pass_emploi_app/models/message.dart';
-import 'package:pass_emploi_app/redux/actions/chat_actions.dart';
-import 'package:pass_emploi_app/redux/states/app_state.dart';
-import 'package:redux/redux.dart';
 
 class ChatRepository {
-  String _collectionPath = '';
-  StreamSubscription<QuerySnapshot>? _messagesSubscription;
-  StreamSubscription<DocumentSnapshot>? _chatStatusSubscription;
-  String? _chatDocumentId;
+  late final String _collectionPath;
 
   ChatRepository(String firebaseEnvironmentPrefix) {
     this._collectionPath = firebaseEnvironmentPrefix + "-chat";
   }
 
-  // TODO unsubscribe depending on app lifecycle
-  subscribeToMessages(String userId, Store<AppState> store) async {
-    unsubscribeToMessages();
-    store.dispatch(ChatLoadingAction());
+  Stream<List<Message>> messagesStream(String userId) async* {
+    final chatDocumentId = await _getChatDocumentId(userId);
+    if (chatDocumentId == null) return;
 
-    final chats =
-        await FirebaseFirestore.instance.collection(_collectionPath).where('jeuneId', isEqualTo: userId).get();
-    _chatDocumentId = chats.docs.first.id;
-
-    final Stream<QuerySnapshot> messageStream = _messagesCollection().orderBy('creationDate').snapshots();
-    _messagesSubscription = messageStream.listen(
-      (QuerySnapshot snapshot) {
-        final messages = snapshot.docs.map((DocumentSnapshot document) => Message.fromJson(document)).toList();
-        store.dispatch(ChatSuccessAction(messages));
-      },
-      onError: (Object error, StackTrace stackTrace) {
-        store.dispatch(ChatFailureAction());
-      },
-      cancelOnError: false,
-    );
+    final Stream<List<Message>> stream = _chatCollection(chatDocumentId)
+        .collection('messages')
+        .orderBy('creationDate')
+        .snapshots()
+        .map((snapshot) => snapshot.docs.map((document) => Message.fromJson(document)).toList())
+        .distinct();
 
-    final Stream<DocumentSnapshot> chatStatusStream = _chatStatusCollection().snapshots();
-    _chatStatusSubscription = chatStatusStream.listen(
-      (DocumentSnapshot snapshot) {
-        final Map<String, dynamic> data = snapshot.data()! as Map<String, dynamic>;
-        final unreadMessageCount = data['newConseillerMessageCount'];
-        final lastConseillerReading = data['lastConseillerReading'];
-        final DateTime? dateTime = lastConseillerReading != null ? (lastConseillerReading as Timestamp).toDate() : null;
-        store.dispatch(ChatConseillerMessageAction(unreadMessageCount, dateTime));
-      },
-      onError: (Object error, StackTrace stackTrace) {
-        print("Chat status error");
-      },
-      cancelOnError: false,
-    );
+    await for (final messages in stream) yield messages;
```

Comment:
Je trouve ça assez dommage d'avoir un stream au dessus et d'être obligé de faire un `yield messages` sur chaque élément, ça aurait été cool de pouvoir faire un `return stream;` tout simplement. 

Mais j'ai regardé rapidement, je pense pas que ça soit possible 🤷 Vous avez essayé aussi ?

Reply:
J'ai creusé un petit peu plus, faire un 
```dart
firestoreStream.listen((messages) async* { yield messages; })
```

à la place du `await for` ça ne marcherait pas ? 🤔

Reply:
De ce que je comprends de [ce lien SO](https://stackoverflow.com/questions/42611880/difference-between-await-for-and-listen-in-dart/42613676) - le `await for` suspend l'exécution  jusqu'à ce que le stream finisse (donc dans notre cas jamais ?) alors que le `listen` pose le listener et finit. Ca serait plus propre avec le `listen`, non ?